src/internal/storage/fileset/storage.go

```diff
@@ -225,17 +226,44 @@ func (s *Storage) getPrimitives(ctx context.Context, ids []ID) ([]*Primitive, er
 // The path ranges must be non-overlapping and the ranges must be lexigraphically sorted.
 // Concat always returns the ID of a primitive fileset.
 func (s *Storage) Concat(ctx context.Context, ids []ID, ttl time.Duration) (*ID, error) {
-	fsw := s.NewWriter(ctx, WithTTL(ttl))
+	var additiveIndexes, deletiveIndexes []*index.Index
+	var size int64
 	for _, id := range ids {
-		fs, err := s.Open(ctx, []ID{id})
+		md, err := s.store.Get(ctx, id)
 		if err != nil {
-			return nil, err
+			return nil, errors.EnsureStack(err)
 		}
-		if err := CopyFiles(ctx, fsw, fs, true); err != nil {
-			return nil, err
+		prim := md.GetPrimitive()
+		if prim == nil {
+			return nil, errors.Errorf("file set %v is not primitive", id)
 		}
+		additiveIndexes = append(additiveIndexes, prim.Additive)
+		deletiveIndexes = append(deletiveIndexes, prim.Deletive)
+		size += prim.SizeBytes
+	}
+	additive, err := s.concat(ctx, additiveIndexes)
+	if err != nil {
+		return nil, err
+	}
+	deletive, err := s.concat(ctx, deletiveIndexes)
+	if err != nil {
+		return nil, err
+	}
+	return s.newPrimitive(ctx, &Primitive{
+		Additive:  additive,
+		Deletive:  deletive,
+		SizeBytes: size,
+	}, ttl)
+}
+
+func (s *Storage) concat(ctx context.Context, indexes []*index.Index) (*index.Index, error) {
+	mergeIndex := index.NewWriter(ctx, s.ChunkStorage(), "index-concat")
+	if err := index.Merge(ctx, s.ChunkStorage(), indexes, func(idx *index.Index) error {
```

Comment:
Just to make sure I'm following, the main functional change here is that the concat step of compaction will now just always concatenate indices, whereas before it might actually do an expensive rewrite of the underlying data?

Reply:
Right. The existing system would, in general, run the chunks at the edges of the shards through the content defined chunking to get the stable split points. This isn't really necessary anymore and the tradeoff for removing it is slight potential for less deduplication but no longer having to move file content through the PFS master.